### DISCUSSION TITLE
DOCKER: Build Docker images ?

### DISCUSSION THREAD

**Discussion by prologic** [modoboa/modoboa#821] [Q&A]
Has anyone tried to create a working set of Docker images for modoboa, postfix, dovecot?

**Comment by hadifarnoud** [comment]
I'm also interested in this. will try to build one though

**Comment by prologic** [comment]
Built a set of images off of the Alpine base image for optimized size and
quicker deployment :)

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by hadifarnoud** [comment]
care to share @prologic ?

**Comment by prologic** [comment]
Ooops; sorry that was a typo! I meant "build"; not that I've already "built"! :) I should sit down and do this though!

**Comment by tonioo** [comment]
@prologic Would be awesome ;-)

**Comment by hadifarnoud** [comment]
I would be glad to help @prologic

**Comment by prologic** [comment]
So I guess the first thing we need to do is build a set of images.

Ideally we need to define clear separations between "services".

I have come across the following images that I think are most likely suitable bases to start from:
- https://github.com/tomav/docker-mailserver
- https://github.com/lava/dockermail
- https://github.com/adaline/dockermail

Bare in mind however that none of these images are optimized for size. I believe most of them use Ubuntu as their base image.

Getting a Docker image for modoboa is probably far easier and could basically be something like this (_untested_):

`Dockerfile`:

``` #!bash
FROM prologic/python-runtime:2.7-onbuild

ENTRYPOINT ["python", "manage.py", "runserver"]
CMD []
```

This will automatically via `ONBUILD` install all the requirements by `pip install requirements.txt` and perform a `pip install .` on the package itself. The entrypoint is probably _not quite_ right; but this is a start :)

**Comment by hadifarnoud** [comment]
setting up modoboa is such a pain. I tried installer and for some reason I was not able to find out, I could not send email. manual install is hard.

therefore, I was not able to create a docker image. you had any luck?

**Comment by prologic** [comment]
I haven't gotten around to this yet; no.

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by prologic** [comment]
Okay going to give this a crack :)

**Comment by prologic** [comment]
Well I got an initial image up and working, but I get an error immediately upon testing http://192.168.99.100:8000/

``` #!python
Environment:


Request Method: GET
Request URL: http://192.168.99.100:8000/

Django Version: 1.5.12
Python Version: 2.7.10
Installed Applications:
('django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.sites',
 'django.contrib.messages',
 'django.contrib.staticfiles')
Installed Middleware:
('django.middleware.common.CommonMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware')


Traceback:
File "/usr/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  101.                     resolver_match = resolver.resolve(request.path_info)
File "/usr/lib/python2.7/site-packages/django/core/urlresolvers.py" in resolve
  338.             for pattern in self.url_patterns:
File "/usr/lib/python2.7/site-packages/django/core/urlresolvers.py" in url_patterns
  366.         patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
File "/usr/lib/python2.7/site-packages/django/core/urlresolvers.py" in urlconf_module
  361.             self._urlconf_module = import_module(self.urlconf_name)
File "/usr/lib/python2.7/site-packages/django/utils/importlib.py" in import_module
  35.     __import__(name)
File "/app/mail/mail/urls.py" in <module>
  4.     url(r'', include('modoboa.urls')),
File "/usr/lib/python2.7/site-packages/django/conf/urls/__init__.py" in include
  25.         urlconf_module = import_module(urlconf_module)
File "/usr/lib/python2.7/site-packages/django/utils/importlib.py" in import_module
  35.     __import__(name)
File "/usr/lib/python2.7/site-packages/modoboa/urls.py" in <module>
  7. load_core_settings()
File "/usr/lib/python2.7/site-packages/modoboa/core/__init__.py" in load_core_settings
  52.     parameters.register(GeneralParametersForm, ugettext_lazy("General"))
File "/usr/lib/python2.7/site-packages/modoboa/lib/parameters.py" in register
  176.     form = formclass()
File "/usr/lib/python2.7/site-packages/modoboa/core/app_settings.py" in __init__
  212.         super(GeneralParametersForm, self).__init__(*args, **kwargs)
File "/usr/lib/python2.7/site-packages/modoboa/lib/parameters.py" in __init__
  56.             self._load_initial_values()
File "/usr/lib/python2.7/site-packages/modoboa/lib/parameters.py" in _load_initial_values
  87.         from .models import Parameter
File "/usr/lib/python2.7/site-packages/modoboa/lib/models.py" in <module>
  17. reversion.register(Parameter)

Exception Type: AttributeError at /
Exception Value: 'module' object has no attribute 'register'
```

Any ideas here?

**Comment by prologic** [comment]
FYI: So far I've managed to get this image to a respectable:

``` #!bash
$ docker images prologic/modoboa
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
prologic/modoboa    latest              423f52d89111        5 minutes ago       257.9 MB
```

Unfortunately I have to pull in `build-base` for [lxml](https://pypi.python.org/pypi/lxml). I have some ideas to further optimize the size and potentially do away with `build-base` and `-dev` packages.

**Comment by tonioo** [comment]
You could try to install lxml from system packages ?
Do you still encounter the issue you pasted ?

**Comment by prologic** [comment]
Let me try that and get back to you :)

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by prologic** [comment]
@tonioo is the current master in a working state?

**Comment by tonioo** [comment]
It should. Why ?

**Comment by prologic** [comment]
I haven't had much luck yet; so I'm going to build this on my OS X laptop
(instead of Docker) and see if I can repro this issue.

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by hadifarnoud** [comment]
this could potentially help me solve setting up modoboa. any progress?

**Comment by prologic** [comment]
I haven't made any more progress in this; no. Sorry. I've been
concentrating on https://github.com/adaline/dockermail that I got up and
running successfully which just so happens to be a nice modular set of
images for an almost complete mail server solution.

I believe modoboa could fit nicely into adaline/dockermail but I haven't
gone back to this yet.

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by pierreozoux** [comment]
As we are already linked here, I can also show my work to have various containers to tackle the mail hosting issue:
- https://github.com/indiehosters/email

I'm currently using vimbadmin, quiete hapoy with it, but as @almereyda said, I want to have ldap integration, and looks like you are providing it.

We'll meet with @almereyda in July, and if I decide to go for modoboa, then I'll spend time to switch my setup, and I'd be happy to share my work.

In the mean time, if you want to try my setup, and have any questions about, feel free, I'd be really happy to answer!

**Comment by tonioo** [comment]
@pierreozoux Thank you for sharing your experience here, really appreciated.

**Comment by thiesschneider** [comment]
any news on this? i am very interested in this!

**Comment by pierreozoux** [comment]
So, we evaluated modoboa, and we though it would write ldap, but it is just read only. So our guess with @almereyda is to develop our own solution. Sorry.

**Comment by hadifarnoud** [comment]
@pierreozoux hopefully you share that?

**Comment by almereyda** [comment]
This needs a more thorough architectural plan, where we need to identify separate services for containerisation. Right now there are already Postfix, Dovecot, Amavis, Nginx, uWSGI, the cronjobs and eventually Radicale known to me. Then we can recompose them to a linked environment as in @pierreozoux's preliminary work.

We don't like monolithic containers which act like VMs, as in https://github.com/tozd/docker-mail

**Comment by almereyda** [comment]
https://github.com/mail-in-a-box/mailinabox/issues/590#issuecomment-269860069 tells us about https://mailcow.email/dockerized/ and its code in https://github.com/andryyy/mailcow-dockerized

Mailu (ex-Freeposte.io) now also has a Docker Compose example in https://github.com/Mailu/Mailu

**Comment by hadifarnoud** [comment]
have you used mailu @almereyda? is it better than modoboa?

**Comment by msenebald** [comment]
Hi, 
i can quickly describe my current solution.
Basically it is a "all in one" solution which uses the installer during the build. 
Not the most efficient image design but self contained. 
State is handled via some basic volumes.
I prepared a supervisord file for startup
rsyslog logs to console so everything is in docker logs.
Nginx in front of container for webinterface

Since I dont have this in public repo yet, i share some snippets.
I agree with @almereyda a more architectural plan is needed. 
My approach was to have a simple container solution. You could easily take out db 
but postfix/dovecot you have a kind of tight integration. That becomes a complex beast pretty easily if you take apart the services. 
I could imagine having a good dockerfile/build process so everyone choose the bits and pieces needed for them. I do that currently by having a install.cfg ready where i choose what extensions I want. Upgrade path is rebuilding image and migrate  modoboa on deployed container.

Dockerfile:

```Dockerfile
FROM debian:jessie
WORKDIR /usr/src/app
RUN set -ex && \
         apt-get update && \
         apt-get install -y supervisor ca-certificates python git wget nano net-tools dnsutils rsyslog && \
         apt-get clean && \
         git clone https://github.com/modoboa/modoboa-installer.git /usr/src/app/ && cd /usr/src/app
COPY installer.cfg .
RUN ./run.py --force hostname --debug && \
        postalias /etc/aliases 
COPY files /
CMD ["supervisord"]
```

Volumes:
```
modoboa_db:/var/lib/postgresql  
modoboa_vmail:/srv/vmail  
modoboa_app:/srv/modoboa  
modoboa_postfix:/var/spool/postfix/  
modoboa_etc:/etc  
```

In any case getting a good docker solution I am willing to help with, as I have it productive running already.

**Comment by jaketame** [comment]
Would need to ensure that it would be good to have both options available ie local install or docker install. If you move from local install to docker then a migration path from local to docker as well.

**Comment by stefaweb** [comment]
Found this Docker project if this can help.

https://code.georgi.software/patrick/docker-containers/tree/master/mail-modoboa

**Comment by ntninja** [comment]
Self-plug: I've been maintaining (and slowly improving) my own Docker build of Modoboa over the last 1,5 years: https://hub.docker.com/r/ntninja/modoboa/

It now comes with basic documentation (quick-start compose setup, running basic command & upgrading) and has separate containers for the following services: modoboa itself [running via gunicorn not uwsgi], Postfix & Dovecot; meaning it's lacking: Amavis (I use rspamd instead), OpenDKIM (again, I use rspamd instead), [AutoMX](https://modoboa.org/en/weblog/2012/12/16/make-users-life-easier-with-automx/), Radicale (the contacts app works but refuses to sync CardDAV without it), ~~the stats plugin is broken since there is no `/var/log/mail.log`~~ fixed the stats plugin by submitting log messages using syslog/udp to modoboa container where they are written to file and automatically collected and deleted every 5 minutes.

**Comment by TildeSlashC0re** [comment]
@alexander255 I've recently forked your `docker-recipes` [gitlab repo ](https://gitlab.com/ntninja/docker-recipies/) and re-structured it to focus exclusively on modoboa. 

What I have covered so far:
* Cleaning up the overall file-structure
* Extending the docker-compose.yml deployment with a persistent postgres-db container
* Added `wait-for-it.sh` to the init process of the modoboa container to wait for postgres to be responsive
  * :exclamation: this is non-trivial when running `deploy`
* Adding ldap-login dependecies
* Compiling all extensions (except modoboa-amavis) at build-time
  * :exclamation: using your proposed way of `docker-compose run modoboa install <extension>` doesn't really cut it since I have to define the utilized extensions at deploy time (== after build, == after gcc gets removed from the apk packages, == guaranteed to fail :smile:)
* Testing deployment with all extensions (but modoboa-amavis) enabled 

Now, as far as I can tell, what's left to do is to extend the docker-compose.yml with the various components to make this thing a viable deployment in the wild. 
E.g.:
* Amavis / spamassasin
* Radicale
* Some webmail suite (?)
* Anything you think I've left out

If you feel like this is a project worth getting on with, let me know and I'll host my local repo on Github / Gitlab so you can take a look and send further suggestions my way.


**P.S:** I'd just like to express my amazement at how this Issue right here is ~3 years old without some dude having dropped a viable solution yet :rofl: 

**Edit:** that [documentation](https://hub.docker.com/r/ntninja/modoboa/) of yours could use some rework (which is a nice way to say it drove me nuts at times :sweat_smile:)

**Comment by hadifarnoud** [comment]
@TildeSlashC0re could this enable a more easy backup-restore as well? 
also, would love to have modoboa on kubernetes. this can help with that too. hopefully a Helm chart in near future 🥇

**Comment by ntninja** [comment]
@TildeSlashC0re:

> @alexander255 I've recently forked your `docker-recipes` [gitlab repo ](https://gitlab.com/ntninja/docker-recipies/) and re-structured it to focus exclusively on modoboa.

Could you upload that somewhere? I'd like to take a look!

>     * Extending the docker-compose.yml deployment with a persistent postgres-db container

You mean the quickstart file? Yes, that sounds like a good idea as you as you highlight in the comments that an external container may be used instead. The current quickstart file is basically a dumbed down version of my server configuration with comments. :wink: 

>     * Added `wait-for-it.sh` to the init process of the modoboa container to wait for postgres to be responsive
>       * this is non-trivial when running `deploy`

:+1: 
 
>     * Adding ldap-login dependecies

I tried to do that everywhere, but I guess I forgot some place, since I don't actually use it myself :slightly_smiling_face: 

>     * Compiling all extensions (except modoboa-amavis) at build-time
>       * using your proposed way of `docker-compose run modoboa install <extension>` doesn't really cut it since I have to define the utilized extensions at deploy time (== after build, == after gcc gets removed from the apk packages, == guaranteed to fail smile)

I disagree on this: The core should be small IMHO (I invested quite some work into reducing size in several places); if you look at the current implementation for instance, it *will* ship `rrdtool` + CPython bindings in the base image for instance since `modoboa-stats` requires that – it will not ship anything that Pure-Python can be installed using `pip` however. I do not see why this shouldn't work for Amavis (or any other extension).

>     * Testing deployment with all extensions (but modoboa-amavis) enabled

:+1: I only tested with the extensions documented on the documentation page.

> Now, as far as I can tell, what's left to do is to extend the docker-compose.yml with the various components to make this thing a viable deployment in the wild.
> E.g.:
> 
>     * Amavis / spamassasin
>     * Radicale
>     * Some webmail suite (?)
>     * Anything you think I've left out

 * I've started working on Amavis some time ago, but never completed it since I ended up using `rspamd` instead. I'll make the build stuff available, since it may be useful. – Help welcome.
 * Radicale similarily I never started since we have been using ownCloud (now NextCloud) for longer than we use Modoboa (lack of contact sync with NextCloud is a pity however) – Any contributions welcome of course as well.
 * Webmail is provided by `modoboa-webmail` and works without issues
 * I'd like to also integrate `automx` for automatic mail client configuration (ideally as a Modoboa plugin), but it's a low-priority project

> If you feel like this is a project worth getting on with, let me know and I'll host my local repo on Github / Gitlab so you can take a look and send further suggestions my way.

I see, yes please do. Are you planning on deploying this in the future and maybe helping out in development / maintenance / documentation / testing / whatever else? If so, would you be interested in creating a group for this? (As you may have guessed I'd prefer using GitLab over GitHub.)

> **P.S:** I'd just like to express my amazement at how this Issue right here is ~3 years old without some dude having dropped a viable solution yet rofl

Several started on it (see above), but no-one completed it. Maybe this time it's different. :wink: 

> **Edit:** that [documentation](https://hub.docker.com/r/ntninja/modoboa/) of yours could use some rework (which is a nice way to say it drove me nuts at times sweat_smile)

Absolutely, it started out as a bunch of notes to myself on how to do upgrades and stuff and most of the “how to deploy” kind of stuff has only been written a couple of weeks ago as I tried to get this into shape for initial publishing (it's been around for a while of course, but IMHO it was pretty worthless without the documentation on how to deploy it). I've added a copy of the Docker repo's README to GIT for easier editing: https://gitlab.com/ntninja/docker-recipies/blob/master/modoboa/README.md

**Comment by ntninja** [comment]
@hadifarnoud:

> could this enable a more easy backup-restore as well?

You only need to back up the data directory, with the mail directory and Docker configuration, and the database. So yes, makes it easier for me. :slightly_smiling_face: 

> also, would love to have modoboa on kubernetes. this can help with that too. hopefully a Helm chart in near future

I wanted to start using Kubernetes for quite some time. If you could provide some relevant configuration for this I'd be very interested.

**Comment by hadifarnoud** [comment]
@alexander255 I by no means am an expert on k8s. I'd give it a try.

**Comment by TildeSlashC0re** [comment]
@alexander255
> 
> Could you upload that somewhere? I'd like to take a look!
> 

Officially forked it on gitlab now: [c0re/modoboa-dockerized](https://gitlab.com/TildeSlashC0re/modoboa-dockerized)
since I didn't push any of the docker images on dockerhub yet you are advised to build them locally (see 
 `build-forked-images.sh` for reference)

> You mean the quickstart file? Yes, that sounds like a good idea as you highlight in the comments that an external container may be used instead. The current quickstart file is basically a dumbed down version of my server configuration with comments. 

yep I've based my project on the work you've provided on the quickstart deployment. 

> I tried to do that everywhere, but I guess I forgot some place, since I don't actually use it myself 

I had to add openldap-dev to the base-packages and django-auth-ldap to the pip packages to get it to work. Since modoboa ships with LDAP integration per default I thought that'd be a valuable add.

Not to mention the on-the-fly LDAP integration is the main reason why I have taken an interest in this suite :smile:  

> > ```
> > * Compiling all extensions (except modoboa-amavis) at build-time
> >   * using your proposed way of `docker-compose run modoboa install <extension>` doesn't really cut it since I have to define the utilized extensions at deploy time (== after build, == after gcc gets removed from the apk packages, == guaranteed to fail smile)
> > ```
> 
> I disagree on this: The core should be small IMHO (I invested quite some work into reducing size in several places); if you look at the current implementation for instance, it _will_ ship `rrdtool` + CPython bindings in the base image for instance since `modoboa-stats` requires that – it will not ship anything that Pure-Python can be installed using `pip` however. I do not see why this shouldn't work for Amavis (or any other extension).

I don't disagree on the presupposition that the base image should remain as small as viably possible, we will agree that this is the common goal with **any** dockerized deployment :smile:. 
The issue I tried to highlight is that the build-dependencies (namely gcc) are being removed as soon as the pip packages are done installing, which makes it impossible (at least with the [Dockerfile provided](https://gitlab.com/TildeSlashC0re/modoboa-dockerized/blob/b74fd4ab43db0d380df38635c1312d4d2c851284/modoboa/Dockerfile)) to implement additional extensions after the image is done building.

I totally could be missing something here, and if so - by all means - let me know. 

As an aside, I've squashed my current build process with the following results (how come your dovecot image (via dockerhub) is twice the size from the [Dockerfile](https://gitlab.com/TildeSlashC0re/modoboa-dockerized/blob/b74fd4ab43db0d380df38635c1312d4d2c851284/alpine-dovecot/Dockerfile) in your `docker-recipes` repo?!) 

(prebuilt-ext tag corresponds to the [Dockerfile.full](https://gitlab.com/TildeSlashC0re/modoboa-dockerized/blob/master/deps/modoboa/Dockerfile.full) build)
```
REPOSITORY                    TAG                     SIZE
c0re/modoboa                  prebuilt-ext-squashed   250MB
c0re/modoboa                  prebuilt-ext            442MB

ntninja/modoboa               latest                  202MB
c0re/modoboa                  squashed                225MB
c0re/modoboa                  latest                  418MB

ntninja/modoboa-postfix       latest                  13.3MB
c0re/modoboa-postfix-alpine   squashed                13.3MB
c0re/modoboa-postfix-alpine   latest                  13.6MB

ntninja/modoboa-dovecot       latest                  61.7MB
c0re/modoboa-dovecot-alpine   squashed                25.5MB
c0re/modoboa-dovecot-alpine   latest                  25.9MB
```

> > ```
> > * Testing deployment with all extensions (but modoboa-amavis) enabled
> > ```
> 
>  I only tested with the extensions documented on the documentation page.

Same goes for me. What I meant to say here is "all extensions" as in:
```
docker-compose run modoboa deploy --collectstatic \
    --domain <fqdn> \ 
    --dburl default:postgres://<user:password>@<host:port>/<database> \
    --extensions all
```

> * I've started working on Amavis some time ago, but never completed it since I ended up using `rspamd` instead. I'll make the build stuff available, since it may be useful. – Help welcome.

aye aye!

> * Radicale similarily I never started since we have been using ownCloud (now NextCloud) for longer than we use Modoboa (lack of contact sync with NextCloud is a pity however) – Any contributions welcome of course as well.

Funny, because I am in the exact same situation here. Mainly using Nextcloud for all calDav operations. 
Do you mean contact sync as in modoboa <> nextcloud? Or a Nextcloud calDav endpoint for contact sync in general? 
(since I know for a fact that you can sync contacts with the Nextcloud calDav)

> * Webmail is provided by `modoboa-webmail` and works without issues

huh... I'll have to look into that... might as well be the local nature of my deployment that didn't make it work. Although I'll admit I haven't spent too much time on it since I also manage webmail mainly through Nextclouds solution.

> * I'd like to also integrate `automx` for automatic mail client configuration (ideally as a Modoboa plugin), but it's a low-priority project

aye aye!

> Are you planning on deploying this in the future and maybe helping out in development / maintenance / documentation / testing / whatever else? If so, would you be interested in creating a group for this? (As you may have guessed I'd prefer using GitLab over GitHub.)

I could definitely spare some time to assist with any of the tasks named here. Not exaclty sure if I can spare enough of it for full maintenance, but we could see about that along the way :+1: 

kind regards, 
c0re

**Comment by TildeSlashC0re** [comment]
I'll announce a little bit of progress on my end here:

So far I have:
* Built a full-featured (that is `deploy --extensions all` capable) **modoboa** container (thanks to the Dockerfile provided by @alexander255) weighing in at **251 MB**
* Built a **postgres** container definition that handles connections to both modoboa and amavis @ **70 MB**
* Built an **amavis** container based on [muhlba91/amavis-docker](https://github.com/muhlba91/amavis-docker)'s Dockerfile
  * This - being debian-based (and poorly stripped (that is to say not at all)) turned out to be a rather heavy deployment, weighing in at **464 MB**. A quick `--no-install-recommends` has cut that just about in half, down to **251 MB** - with all possible stripping and squashing the final piece sat at **242 MB**. 
  * Now I am a curious cookie and was eager to see what the competition is bringing to the table and lo' and behold... mailcow-rspamd ships at around **213 MB**. I am a perfectionist like that and thought to myself: fudge it, time to alpine-ize it. 
Long story short: My current non-polished (but squashed) local dev-build for modoboa-amavis-alpine comes at the low cost of **183 MB** :smirk: 
* Made those pieces play together via a `docker-compose.yml` (again, the major basis for this is by @alexander255)
  * I will eventually upgrade this definition to a v3 format. 

So up to this stage every moving piece is built on an alpine base... and I will certainly continue on this route.
The savings in space/bandwidth are just completely out there!

## ToDo:
* A bit of permission auditing across the various mounted data-volumes. (PDFcreds-export wasn't really playing nice when I tried running it the first time around)
* Provide ManageSieve compatibility
* Add a **radicale** instance into the mix
* Implement **autoMX** capabilities
* Adjust the postgres image to host an additional database for **spamassassin**
* Make some final tweaks for **full** quarantine / release-request / PDP integration. (Basically deploying and running the full scope of the amavis integration)
* Provide a streamlined / guided installation procedure
* Provide documentation
* Tests... lots of them
* Anything I have either forgotten about or have simply overlooked. (any feedback is welcome!)

kind regards,
c0re :v:

**Comment by Zegorax** [comment]
@TildeSlashC0re Is it possible to change c0re/modoboa, c0re/modoboa-dovecot-alpine and c0re/modoboa-postfix-alpine to public on the Docker Hub ?

Thanks for your work !

**Comment by TildeSlashC0re** [comment]
> @TildeSlashC0re Is it possible to change c0re/modoboa, c0re/modoboa-dovecot-alpine and c0re/modoboa-postfix-alpine to public on the Docker Hub ?
> 
> Thanks for your work !

I will provide public images once this takes a bit more of a final shape.
Up until then you are advised to build the images locally, all of which are provided [here](https://gitlab.com/TildeSlashC0re/modoboa-dockerized/tree/devel/deps). Alternatively you can use the [helper script](https://gitlab.com/TildeSlashC0re/modoboa-dockerized/blob/devel/build-forked-images.sh) to build the images. 

The first release will include an introductory documentation and public images. :v:

**Comment by Zegorax** [comment]
Ahh perfect ! I was searching for the Dockefiles but I don't know how I could have missed them !

Thanks ! Will try it right away. I'll open issues / give you my feedback

**Comment by TildeSlashC0re** [comment]
I might add that the amavis build currently isn't baked into the docker-compose and that there still is quite some manual configuration ahead on an initial deployment. 
All of which is undocumented as of now. 

I'll gladly answer any issues and will look to build a basic documentation along the way.

**Comment by Zegorax** [comment]
For the moment when I start everything up I get a 400 Bad Request from Modoboa. Do you know the workaround maybe ?

**Comment by TildeSlashC0re** [comment]
First run is something like this atm:
Initialize the database
```
docker-compose up db
```
You can exit the container when the setup scripts are executed. 
Then you deploy modoboa itself.

```
docker-compose run modoboa deploy --collectstatic --domain "EXAMPLE.COM" --dburl default:postgres://modoboa:mysecretpassword@db:5432/modoboa --extensions all
```

`mysecretpassword` corresponds to the value you have specified for `POSTGRES_PASSWORD` in the docker-compose.yml

after the deployment you need to edit the config file `configs/settings.py` and add a database-block for amavis:

```py
[...]

# Database
# https://docs.djangoproject.com/en/1.11/ref/settings/#databases

DATABASES = {
    
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'modoboa',
        'USER': 'modoboa',
        'PASSWORD': 'mysecretpassword',
        'HOST': 'db',
        'PORT': '5432',
        'ATOMIC_REQUESTS': True,

    },

    'amavis': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'amavis',
        'USER': 'amavis',
        'PASSWORD': 'mysecretpassword',
        'HOST': 'db',
        'PORT': '5432',
        'ATOMIC_REQUESTS': True,

    },

}

[...]
```

then start amavis:
```
docker run --rm --network modoboa-dockerized_default -v $(pwd)/data/amavis:/data -d --hostname "host.example.com" --name modoboa-amavis c0re/modoboa-amavis-alpine
```

You might want to change settings in `deps/modoboa-amavis-alpine-s6/rootfs/etc/amavisd.conf ` (such as `$mydomain`) and then rebuild the image `cd deps/modoboa-amavis-alpine-s6 && docker build -t c0re/modoboa-amavis-alpine .`

you should then be ready to start modoboa and configure the settings in the webpanel

```
docker-compose up -d
```

**Comment by Zegorax** [comment]
Perfect thank you very much !
FYI : It is not necessary to edit amavisd.conf. When running Amavis, simply add a hostname parameters to the docker container and Amavis will use it.

Example to add to your docker-compose.yml : 

```
amavis:
    image: c0re/modoboa-amavis-alpine
    hostname: example.com
    volumes:
      - ./data/amavis:/data
```

**Comment by Zegorax** [comment]
The last problem present is this one with dovecot : 

`modoboa-dovecot | doveconf: Fatal: Error in configuration file /etc/dovecot/conf.d/10-ssl.conf line 50: ssl_dh: Can't open file /etc/dovecot-local/dh.pem: No such file or directory
modoboa-dovecot exited with code 89
modoboa-dovecot | Keeping existing /etc/certs/fullchain.pem
modoboa-dovecot | doveconf: Fatal: Error in configuration file /etc/dovecot/conf.d/10-ssl.conf line 50: ssl_dh: Can't open file /etc/dovecot-local/dh.pem: No such file or directory`

I think basic certificates are not generated in the image build

**Comment by TildeSlashC0re** [comment]
ah yes! You need to supply the DH-parameters yourself. You can generate those by running 
`openssl dhparam -out ./data/dovecot/config/dh.pem 4096`

**Comment by Dave-Elec** [comment]
I was trying to run this docker and everything started fine. But I am getting csrf verification failed. I mapped 80:8000 to get the webmail for trial. I am going to use nginx later. I think it is an issue with the webmail only. What should I do? Thanks for the docker setup.
